### PR TITLE
Indexer: Escape special characters in storyImport regex

### PR DESCRIPTION
**Pull request by vojgin**
Closes #

No issue in evidence

## What I did

I had a problem with using `<Meta of{...} />` inside MDX file.

238134010-5c70d6c1-8dec-40ae-8e93-fc7a27144c3c

It always resulted in this error and I didn't know why.

Turns out the problem was the string literal inserting in the Regex as it did not escape special characters from the file path thus it could never match.

Usually paths have slashes `/` which work fine without escaping. But I had a plus sign `+` in my path. A plus sign is regex modifier, thus the regex stopped working.

## How to test

1. Run **unfixed** storybook on a project with plus sign `+` somewhere in the absolute path and try to include some MDX containing `<Meta of={...} />`

Such as this one:

```mdx
{/* ButtonDocs.mdx */}

import { Meta, Story } from '@storybook/blocks';
import * as ButtonStories from './Button.stories';

<Meta of={ButtonStories} />
```

2. **It will result in an error.**

3. You can compare it then with the fixed version then. The error is gone.

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Comment by shilman** [comment]
Also probably related https://github.com/storybookjs/storybook/pull/22110

File: code/lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 1]
```diff
@@ -257,7 +257,9 @@ export class StoryIndexGenerator {
           if (!cacheEntry || cacheEntry.type !== 'stories') return false;
 
           return !!absoluteImports.find((storyImport) =>
-            fileName.match(new RegExp(`^${storyImport}(\\.[^.]+)?$`))
+            fileName.match(
+              new RegExp(`^${storyImport.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}(\\.[^.]+)?$`)
+            )
```
Comment (ndelangen) [inline code/lib/core-server/src/utils/StoryIndexGenerator.ts:262]: I think this is the only significant change? the rest is more cleanup/linting?

**Review by JReinhold — APPROVED** [review]
I think this change looks good, GPT-4 conjured up the exact same string replacement, which obviously means this is pristine quality! 💅

**Comment by JReinhold** [comment]
Most likely also fixes https://github.com/storybookjs/storybook/issues/25114

**Pull request merged**